### PR TITLE
fix: set queued-build-hook state directory

### LINKED ISSUE #9: set HOME for systemd service
I just started using `queued-build-hook` and I love it -- thank you for making and maintaining it!

Unfortunately, it took quite a bit of time to figure out how to set it up correctly (using `test.nix` as an example and reading through `module.nix`).

When the service would attempt to upload things, I'd see errors like these in the logs constantly:

```console
Jun 16 04:53:13 nixos-desktop async-nix-post-build-hook-start[4531]: Uploading paths /nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc
Jun 16 04:53:13 nixos-desktop async-nix-post-build-hook-start[4589]: Uploading paths /nix/store/rx3pwvj59sbd5hvda5hg3prln7h47rsq-nixos-system-nixos-desktop-23.11.20230612.0eeebd6
Jun 16 04:53:19 nixos-desktop async-nix-post-build-hook-start[4531]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:19 nixos-desktop async-nix-post-build-hook-start[4589]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:24 nixos-desktop async-nix-post-build-hook-start[4793]: Uploading paths /nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc
Jun 16 04:53:24 nixos-desktop async-nix-post-build-hook-start[4814]: Uploading paths /nix/store/rx3pwvj59sbd5hvda5hg3prln7h47rsq-nixos-system-nixos-desktop-23.11.20230612.0eeebd6
Jun 16 04:53:31 nixos-desktop async-nix-post-build-hook-start[4793]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:31 nixos-desktop async-nix-post-build-hook-start[4814]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:36 nixos-desktop async-nix-post-build-hook-start[4837]: Uploading paths /nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc
Jun 16 04:53:36 nixos-desktop async-nix-post-build-hook-start[4842]: Uploading paths /nix/store/rx3pwvj59sbd5hvda5hg3prln7h47rsq-nixos-system-nixos-desktop-23.11.20230612.0eeebd6
Jun 16 04:53:42 nixos-desktop async-nix-post-build-hook-start[4837]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:42 nixos-desktop async-nix-post-build-hook-start[4842]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:47 nixos-desktop async-nix-post-build-hook-start[4882]: Uploading paths /nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc
Jun 16 04:53:47 nixos-desktop async-nix-post-build-hook-start[4887]: Uploading paths /nix/store/rx3pwvj59sbd5hvda5hg3prln7h47rsq-nixos-system-nixos-desktop-23.11.20230612.0eeebd6
Jun 16 04:53:53 nixos-desktop async-nix-post-build-hook-start[4882]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:53 nixos-desktop async-nix-post-build-hook-start[4887]: error: creating directory '//.cache': Read-only file system
Jun 16 04:53:58 nixos-desktop async-nix-post-build-hook-start[4934]: Uploading paths /nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc
Jun 16 04:53:58 nixos-desktop async-nix-post-build-hook-start[4939]: Uploading paths /nix/store/rx3pwvj59sbd5hvda5hg3prln7h47rsq-nixos-system-nixos-desktop-23.11.20230612.0eeebd6
Jun 16 04:54:04 nixos-desktop async-nix-post-build-hook-start[4934]: error: creating directory '//.cache': Read-only file system
Jun 16 04:54:04 nixos-desktop async-nix-post-build-hook-start[4939]: error: creating directory '//.cache': Read-only file system
Jun 16 04:54:09 nixos-desktop async-nix-post-build-hook-start[4098]: queued-build-hook: Dropped message with DRV_PATH '/nix/store/3vplfqnjv9jhsb6yb2843fxwpwk6681d-etc.drv' with OUT_PATHS '/nix/store/94w1rm0b84sjhdld2x2mvm27z22p9bpb-etc' afte>
```

I fixed it by manually setting the `XDG_CACHE_HOME` variable:
https://github.com/ConnorBaker/nixos-configs/commit/b2677760e7c5d8682a97b2828c46db1ca0c75e27

Is there a better way to fix this?

I saw a similar bug report on WSLU (https://bugs.launchpad.net/ubuntu/+source/wslu/+bug/1859509), which was caused by the `$HOME` being unset or `/`.

### PR DESCRIPTION
**Pull request by jfroche**
fixes #9

### CODE REVIEW

File: module.nix [hunk 1]
```diff
@@ -159,6 +172,7 @@ in
             KillMode = "process";
             Restart = "on-failure";
             FileDescriptorStoreMax = 1;
+            StateDirectory = baseNameOf cfg.stateDirectory;
```
Comment (Mic92) [inline module.nix:175]: StateDirectory always have be in /var/lib. I'd suggest to make stateDirectory not configurable until we have a valid use case to make it configurable.

File: module.nix [hunk 2]
```diff
@@ -111,6 +119,10 @@ in
       in
       "${enqueueScript}/bin/enqueue-package";
 
+    systemd.tmpfiles.rules = [
+      "d '${cfg.stateDirectory}' - queued-build-hook queued-build-hook - -"
+    ];
```
Comment (Mic92) [inline module.nix:124]:
```suggestion
```

File: module.nix [hunk 3]
```diff
@@ -24,6 +24,14 @@ in
       type = types.path;
     };
 
+    stateDirectory = mkOption {
+      description = mdDoc ''
+        Path to the queued-build-hook state directory.
+      '';
+      default = "/var/lib/async-nix-post-build-hook";
+      type = types.path;
+    };
```
Comment (Mic92) [inline module.nix:33]:
```suggestion
```

File: module.nix (hunk 1 above)
Comment (Mic92) [inline module.nix:175]:
```suggestion
            StateDirectory = "async-nix-post-build-hook";
```

File: module.nix [hunk 4]
```diff
@@ -151,6 +163,7 @@ in
             fi
             exec ${cfg.package}/bin/queued-build-hook daemon --hook ${hook} --retry-interval ${toString cfg.retryInterval} --retry-interval ${toString cfg.retries} --concurrency ${toString cfg.concurrency} 
           '';
+          environment.HOME = cfg.stateDirectory;
```
Comment (Mic92) [inline module.nix:166]:
```suggestion
          environment.HOME = "/var/lib/async-nix-post-build-hook";
```

**Comment by JulienMalka** [comment]
Seems that this PR is in need of love, I am interested in getting it merged as we are hitting #9 in https://github.com/JulienMalka/nix-hash-collection/issues/1, can I help getting it in a mergeable state ?

**Comment by jfroche** [comment]
@Mic92 I have simplified the PR, when you have some time, can you check this once again ? Thank you !

**Pull request merged**